### ISSUE TITLE
Sandbox is slow

### ISSUE THREAD

**Issue by qcoumes** [plgitlogin/sandbox#3]
We have to figure out what is slowing the execution of code on the sandbox.

Here a summary of the time taken by each steps (average of 5 times) from the reception of the request on the sandbox with the parameters:
```python
DOCKER_MEM_LIMIT = "20m"
DOCKER_MEMSWAP_LIMIT = "30m"
DOCKER_CPUSET_CPUS = "0-1"
```

Used builder was `/builder/before.py`
For building:
* `views.BuildView.post()` *from start to Builder call* - `0.004752715` secs
* `executor.Executor.CREATE_DOCKER()` - `0.748968045` secs
* `executor.Executor.move_env_to_docker()` - `0.136438052` secs
* `executor.Builder.build()` - `0.163393815` secs
* `executor.Executor.get_stderr()` - `0.067297618` secs
* `executor.Executor.get_context()` - `0.061285178` secs
***Total:*** `1.182135423` secs

Used grader was `/grader/evaluator.py`
For evaluating:
* `views.EvalView.post()` *from start to Evaluator call* - `0.004553715` secs
* `executor.Executor.CREATE_DOCKER()` - `0.738528645` secs
* `executor.Evaluator.add_answer_to_env()`  - `0.002467473` secs
* `executor.Executor.move_env_to_docker()` - `0.132447051` secs
* `executor.Evaluator.evaluate()` - `0.296804508` secs
* `executor.Executor.get_stderr()` - `0.066254628` secs
* `executor.Executor.get_context()` - `0.067092615` secs
***Total:*** `1,308148635` secs

**Comment by qcoumes** [comment]
Times seems reasonable, we should try to speed up the `executor.Executor.CREATE_DOCKER()` function if it's possible

**Comment by nimdanor** [comment]
The thing to do is to have a pool of dockers.
But this is not the top priority.

**Comment by qcoumes** [comment]
I've worked a bit on it, I managed to lower the time by a factor of 10:

```
[2018/11/24 19:09:10] DEBUG -- Acquiring container 'c0' of id '0'
[2018/11/24 19:09:10] DEBUG -- Acquiring a docker took 0.0002541542053222656
[2018/11/24 19:09:10] DEBUG -- POST BUILD took 0.0008955001831054688
[2018/11/24 19:09:10] DEBUG -- move_env_to_docker() took 0.0014193058013916016
[2018/11/24 19:09:10] DEBUG -- Execution of build with parameters DOCKER_MEM_LIMIT=100m and DOCKER_CPUSET_CPUS=0-4 took 0.1257915496826172 secondes.
0 b''
[2018/11/24 19:09:10] DEBUG -- get_file() took 8.440017700195312e-05
[2018/11/24 19:09:10] DEBUG -- get_context() took 8.678436279296875e-05
[2018/11/24 19:09:10] DEBUG -- Total build took 0.12939858436584473
[2018/11/24 19:09:11] DEBUG -- Releasing container 'c0' of id '0'

```

The code is still a bit rough on the edges, and I havn't upgrade the grader part yet, but it seems promising.

The downside is that starting the server is longer (around 0.7sec per container in  the pool) which we don't really care.

For instance, current computer on which PL is deployed would be able to withstand around 400 requests in around 0.14 secs (the server would take around 30 sec to start tho). With some code on Premier Langage side blocking user from making more than one sandbox request every 0.5 / 1 sec, I think we would have no more problem with the sandbox crashing for being out of memory.



@nimdanor @nthiery @plgitlogin

**Comment by qcoumes** [comment]
Avec les même que les paramètres de l'exemple du premier poste:
```
[2018/11/24 19:52:25] INFO -- Build request received from '127.0.0.1'
[2018/11/24 19:52:25] DEBUG -- Acquiring container 'c0' of id '0'
[2018/11/24 19:52:25] DEBUG -- Acquiring a docker took 0.0002243518829345703
[2018/11/24 19:52:25] DEBUG -- POST BUILD took 0.0008237361907958984
[2018/11/24 19:52:25] DEBUG -- move_env_to_docker() took 0.0013608932495117188
[2018/11/24 19:52:25] DEBUG -- Execution of build with parameters DOCKER_MEM_LIMIT=30m and DOCKER_CPUSET_CPUS=0-1 took 0.13082242012023926 secondes.
[2018/11/24 19:52:25] DEBUG -- get_file() took 7.033348083496094e-05
[2018/11/24 19:52:25] DEBUG -- get_context() took 7.557868957519531e-05
[2018/11/24 19:52:25] DEBUG -- Total build took 0.13418030738830566
[2018/11/24 19:52:25] DEBUG -- Releasing container 'c0' of id '0'
```

**Comment by nthiery** [comment]
> The downside is that starting the server is longer (around 0.7sec per container in the pool) which we don't really care.

And if at the end we did care, nothing should prevent from starting with an empty pool, and filling it in a separate thread, in parallel with serving the first requests, right?

**Comment by qcoumes** [comment]
Solved on 1.0.0
